source/lac/petsc_parallel_sparse_matrix.cc

```diff
@@ -890,6 +890,48 @@ namespace PETScWrappers
       return u*tmp;
     }
 
+    IndexSet
+    SparseMatrix::locally_owned_domain_indices () const
+    {
+      PetscInt n_rows, n_cols, min, max;
+      PetscErrorCode ierr;
+      IS *rows = nullptr;
+      IS *cols = nullptr;
+
+      ierr = MatGetSize (matrix, &n_rows, &n_cols);
+      AssertThrow (ierr == 0, ExcPETScError(ierr));
+      ierr = MatGetOwnershipIS(matrix, rows, cols);
+      AssertThrow (ierr == 0, ExcPETScError(ierr));
+      ierr = ISGetMinMax(*rows, &min, &max);
+      AssertThrow (ierr == 0, ExcPETScError(ierr));
+
+      IndexSet locally_owned_domain_indices(n_rows);
+      locally_owned_domain_indices.add_range(min, max);
```

Comment:
that was exactly my perception that PETSc's IS is quite general and in principle supports non-contiguous sets.

@ESeNonFossiIo : could you please add the extra check as outlined by @bangerth  here and below?